python/hail/typecheck/check.py

```diff
@@ -0,0 +1,202 @@
+
+        self.checkers = flat_checkers
+        super(MultipleTypeChecker, self).__init__()
+
+    def check(self, x):
+        any_pass = False
+        for c in self.checkers:
+            any_pass = any_pass or c.check(x)
+        return any_pass
+
+    def expects(self):
+        return '(' + ' or '.join([c.expects() for c in self.checkers]) + ')'
+
+
+class CollectionChecker(TypeChecker):
+    def __init__(self, collection_checker, element_checker):
+        self.cc = collection_checker
+        self.ec = element_checker
+        super(CollectionChecker, self).__init__()
+
+    def check(self, x):
+        passes = self.cc.check(x)
+        if passes:
+            for elt in x:
+                passes = passes and self.ec.check(elt)
+        return passes
+
+    def expects(self):
+        return '%s[%s]' % (self.cc.expects(), self.ec.expects())
+
+
+class DictChecker(TypeChecker):
+    def __init__(self, key_checker, value_checker):
+        self.kc = key_checker
+        self.vc = value_checker
+        super(DictChecker, self).__init__()
+
+    def check(self, x):
+        passes = isinstance(x, dict)
+        if passes:
+            for k, v in x.iteritems():
+                passes = passes and self.kc.check(k)
+                passes = passes and self.vc.check(v)
+        return passes
+
+    def expects(self):
+        return 'dict[%s, %s]' % (self.kc.expects(), self.vc.expects())
+
+
+def only(t):
+    if isinstance(t, type) or type(t) is ClassType:
+        return LambdaChecker(lambda x: isinstance(x, t), extract(t))
+    elif isinstance(t, TypeChecker):
+        return t
+    else:
+        raise RuntimeError("invalid typecheck signature: expected 'type' or 'TypeChecker', found '%s'" % type(t))
+
+
+def oneof(*args):
+    return MultipleTypeChecker([only(x) for x in args])
+
+
+def nullable(t):
+    return oneof(t, NoneType)
+
+
+def listof(t):
+    return CollectionChecker(only(list), only(t))
+
+
+def dictof(k, v):
+    return DictChecker(only(k), only(v))
+
+
+none = only(NoneType)
+
+anytype = LambdaChecker(lambda x: True, 'Any')
+
+strlike = oneof(str, unicode)
+
+integral = oneof(int, long)
+
+numeric = oneof(int, long, float)
+
+
+def typecheck(*targs, **kw):
+    if len(targs) != 0 and len(kw) != 0:
+        raise RuntimeError('cannot mix positional and keyword type checking. Name all typecheck arguments or none.')
+
+    use_kw = len(kw) != 0
+
+    def get_type(name, index):
+        if use_kw:
+            return kw[name]
+        else:
+            return targs[index]
+
+    def _typecheck(f, *args, **kwargs):
+        argspec = getargspec(f)
+        argspec_args = argspec.args
+
+        self_cls = None
+
+        # strip the 'self' parameter, since this does not need to be checked
+        if len(argspec_args) > 0 and argspec_args[0] == 'self':
+            # try to get more information about the class to write better error messages
+            try:
+                self_cls = args[0].__class__
```

Comment:
This can only be wrong in one of two cases:

1. somebody defines a method that's NOT an instance method with a self parameter:

```
def f(self, x, y):
    return x + y

f(1,2,3)
```

2. Somebody uses an instance method using the expanded syntax, without passing in the correct object:

```
class Foo:
    def bar(self, x, y):
        return 5 + x + y

Foo.bar(1,2,3)
```